src/libpmemfile-core/file.c

```diff
@@ -1674,3 +1674,171 @@ pmemfile_truncate(PMEMfilepool *pfp, const char *path, off_t length)
 
 	return 0;
 }
+
+static int
+vinode_chown(PMEMfilepool *pfp, struct pmemfile_cred *cred,
+		struct pmemfile_vinode *vinode, uid_t owner, gid_t group)
+{
+	struct pmemfile_inode *inode = vinode->inode;
+	int error = 0;
+
+	if (owner == (uid_t)-1 && group == (gid_t)-1)
+		return 0;
+
+	os_rwlock_wrlock(&vinode->rwlock);
+
+	if (!(cred->caps & (1 << PMEMFILE_CAP_CHOWN))) {
+		if (inode->uid != cred->fsuid) {
+			error = EPERM;
+			goto end;
+		}
+
+		if (owner != (uid_t)-1 && owner != inode->uid) {
+			error = EPERM;
+			goto end;
+		}
+
+		if (group != (gid_t)-1 && group != inode->gid) {
+			if (group != cred->fsgid && !gid_in_list(cred, group)) {
+				error = EPERM;
+				goto end;
+			}
+		}
+	}
+
+	COMPILE_ERROR_ON(offsetof(struct pmemfile_inode, gid) !=
```

Comment:
But this assertion might actually better be right next to the pmemobj_tx_add_range_direct call, to make it even more obvious why it is needed. It doesn't do anything runtime anyways.